core/src/apps/cardano/certificates.py

```diff
@@ -0,0 +1,272 @@
+    from typing import List, Tuple, Optional
+    from . import seed
+
+
+POOL_HASH_SIZE = 28
+VRF_KEY_HASH_SIZE = 32
+POOL_METADATA_HASH_SIZE = 32
+PUBLIC_KEY_HASH_SIZE = 28
+IPV4_ADDRESS_SIZE = 4
+IPV6_ADDRESS_SIZE = 16
+
+MAX_URL_LENGTH = 64
+MAX_PORT_NUMBER = 65535
+
+
+def validate_certificate(
+    certificate: CardanoTxCertificateType, protocol_magic: int, network_id: int
+) -> None:
+    if certificate.type in [
+        CardanoCertificateType.STAKE_DELEGATION,
+        CardanoCertificateType.STAKE_REGISTRATION,
+        CardanoCertificateType.STAKE_DEREGISTRATION,
+    ]:
+        if not SCHEMA_STAKING.match(certificate.path):
+            raise INVALID_CERTIFICATE
+
+    if certificate.type == CardanoCertificateType.STAKE_DELEGATION:
+        if not certificate.pool or len(certificate.pool) != POOL_HASH_SIZE:
+            raise INVALID_CERTIFICATE
+
+    if certificate.type == CardanoCertificateType.STAKE_POOL_REGISTRATION:
+        _validate_pool_parameters(
+            certificate.pool_parameters, protocol_magic, network_id
+        )
+
+
+def cborize_certificate(
+    keychain: seed.Keychain, certificate: CardanoTxCertificateType
+) -> Tuple:
+    if certificate.type in [
+        CardanoCertificateType.STAKE_REGISTRATION,
+        CardanoCertificateType.STAKE_DEREGISTRATION,
+    ]:
+        return (
+            certificate.type,
+            [0, get_public_key_hash(keychain, certificate.path)],
+        )
+    elif certificate.type == CardanoCertificateType.STAKE_DELEGATION:
+        return (
+            certificate.type,
+            [0, get_public_key_hash(keychain, certificate.path)],
+            certificate.pool,
+        )
+    elif certificate.type == CardanoCertificateType.STAKE_POOL_REGISTRATION:
+        pool_parameters = certificate.pool_parameters
+
+        return (
+            certificate.type,
+            pool_parameters.pool_id,
+            pool_parameters.vrf_key_hash,
+            pool_parameters.pledge,
+            pool_parameters.cost,
+            cbor.Tagged(
+                30,
+                (
+                    pool_parameters.margin_numerator,
+                    pool_parameters.margin_denominator,
+                ),
+            ),
+            # this relies on pool_parameters.reward_account being validated beforehand
+            # in _validate_pool_parameters
+            get_address_bytes_unsafe(pool_parameters.reward_account),
+            _cborize_pool_owners(keychain, pool_parameters.owners),
+            _cborize_pool_relays(pool_parameters.relays),
+            _cborize_pool_metadata(pool_parameters.metadata),
+        )
+    else:
+        raise INVALID_CERTIFICATE
+
+
+def _validate_pool_parameters(
+    pool_parameters: CardanoPoolParametersType, protocol_magic: int, network_id: int
+) -> None:
+    if any(
+        [
+            not pool_parameters.pool_id
+            or len(pool_parameters.pool_id) != POOL_HASH_SIZE,
```

Comment:
`any()` does not short-circuit, meaning that if pool_id was None, it would result in a runtime error, I tried the following:

```
x = None
any([len(x) > 0])
```

and this fails with `TypeError: object of type 'NoneType' has no len()`

Reply:
right, i did not realize that. it doesn't matter with the `assert_param()` wording though